File: sem3/dsa/sorting/merge_sort.c
#include <stdio.h>

void merge(int array[], int aux[], int mid, int low, int high){
  // Takes an array, which is sorted from low to mid and mid+1 to high
  // This function merges these two halves together to get an ordered array

  // Copy operation
  for (int k=low; k<=high; k++)
    aux[k] = array[k];

  int i = low, j = mid+1;
  for (int k=low; k<=high; k++)
    if (i > mid) array[k] = aux[j++];
    else if (j > high) array[k] = aux[i++];
    else if (aux[i] < aux[j]) array[k] = aux[i++];
    else array[k] = aux[j++];
}

void mergesort(int array[], int aux[], int low, int high){
  if (low >= high) return;

  int mid = (int) low + (high - low) / 2;

  mergesort(array, aux, low, mid);
  mergesort(array, aux, mid+1, high);
  merge(array, aux, mid, low, high);
}

void sort(int array[], int N){
  int aux[N];
  mergesort(array, aux, 0, N-1);
}

int main(){
  int array[] = {1,2,3,9,0,2,5,3,0,6};
  int N = sizeof(array)/sizeof(int);

  sort(array, N);
  for (int i=0; i<N; i++){
    printf("%d ", array[i]);
  }
}

File: sem3/dsa/sorting/merge_sort.py
def merge(array, aux, mid, low, high):
    """Assumes array is sorted from 0 to mid and mid+1 to N"""
    for i in range(low, high+1): # Copy operation
        aux[i] = array[i]
    # i points to A's first, j to B's first and K to aux array's first
    i, j = low, mid+1
    A = array[low:mid]
    B = array[mid+1:high+1]
    for k in range(low, high+1):
        if i > mid:
            array[k] = aux[j]
            j += 1
            continue
        elif j > high:
            array[k] = aux[i]
            i+=1
            continue
        else:
            if aux[i] < aux[j]:
                array[k] = aux[i]
                i+=1
            else:
                array[k] = aux[j]
                j+=1
    return aux

def mergesort(array, aux, low, high):
    if low >= high:
        return
    mid = int(low + (high-low)/2)
    mergesort(array, aux, low, mid)
    mergesort(array,aux, mid+1, high)
    merge(array, aux, mid, low, high)

def sort(array):
    aux = array[:]
    mergesort(array, aux, 0, len(array)-1)
    return array

print(sort([4,2,0,5,2,9,1,0, 3]))
